index.js

```diff
@@ -1,77 +1,72 @@
+    const options = Object.assign({}, params, { hotelIds: id });
+    return Promise.promisify(api_request)(options)
+      .then((results) => {
+        if (!results || results.length === 0) {
+          AwsHelper.log.trace({ hotelId: id }, 'No packages found');
+          return;
         }
-      }
-    }, 500).unref();
-  }
+        AwsHelper.log.trace({ hotelId: id, packageCount: results.length }, 'Found packages');
+        return sendResultsToClient(options, results);
+      });
+  }, { concurrency: 10 });
+}
 
-  api_request(params, function (err, response) { // get packages from NE API
-    var body = JSON.parse(JSON.stringify(params));
-    delete body.hotelIds; // don't need hotelIds again https://git.io/voIAS
-    body.items = []; // send an empty array to the client so it knows wazzup!
-    body.searchComplete = true;
-    AwsHelper.pushResultToClient(body, () => {
-      AwsHelper.log.info({ err: err, packages: response.result.length },
-        'Package search complete');
-    });
-    if (err || !response.result || response.result.length === 0) {
-      AwsHelper.log.warn({ packages: response.result.length }, 'No Packages Found');
-      return callback();
-    } else {
-      return callback(err, response.result.length);
-    }
-  }).on('result', function (body) {
-    resultsReturned++;
-    delete body.hotelIds; // don't need hotelIds again https://git.io/voIAS
-    body.items = body.items.map(function (item) { // so update the list of items
-      item.url = params.searchId + '/' + item.id; // to include an item.url
-      return item;
-    });
-    body.items = body.items.map(mapper.minimiseBandwidth); // minimal fields
-    const id = body.items.length ? body.items[0].id : null;
-    AwsHelper.log.trace({ result_id: id }, 'Sending dynamic package result');
-    AwsHelper.pushResultToClient(body, function (err, data) {
-      /* istanbul ignore if */
-      if (err) {
-        return AwsHelper.log.error({ err: err }, 'Error pushing results to client');
-      }
-      AwsHelper.log.trace({ result_id: id }, 'Sent dynamic package result');
-    });
+function sendResultsToClient (params, results) {
+  return Promise.map(results, (result) => {
+    result.url = `${params.searchId}/${result.id}`;
+    const output = Object.assign(params, { items: [result] });
+    AwsHelper.log.trace({ result: output }, 'Sending result to client');
+    return Promise.promisify(AwsHelper.pushResultToClient)(output)
+      // resolve with data rather than AWS response
+      .then(() => output);
   });
+}
+
+function handler (event, context, callback) {
+  AwsHelper.init(context, event); // to extract the version (ci/prod) from Arn
+  AwsHelper.Logger('lambda-ne-dynamic-package-provider');
+
+  return Promise.resolve()
+    .then(() => {
+      return normaliseParameters(event);
+    })
+    .then((params) => {
+      return searchForPackages(params);
+    })
+    .then((results) => {
+      return results.filter(r => r);
```

Comment:
Strips the empty entries from the results.

If a hotel has no availability then it will create an empty entry in the array. We're not interested in keeping these in the results so filter them out.